### ISSUE TITLE
Better middleware typing

### ISSUE THREAD

**Issue by RomanValihura** [marblejs/marble#78]
First of all, want to say that you doing an awesome job by creating marble.

I have a proposal by better request flow typing.

**Let's consider an example:**
We have an effect middleware for body validation

```typescript
const foo$ = EffectFactory
  .matchPath('/foo')
  .matchType('GET')
  .use(req$ => req$.pipe(
    use(validator$({
      body: Joi.object({
        id: Joi.number().min(1).max(10),
      })
    }));
  ))
     .map(req => ({ body: req.body.id }));
```
After validation we stiil have body as `any`. But we would say that body have type

```typescript
type FooBodyT = { id: number }
```
All the same with any other middlewares that modify request property with `use`. When we will modify some Observable with `map` we will get new type after `map` operator, but `use` not affect the type anyway.

I found useful FP helpers in marble
 * marble/packages/core/src/+internal/fp/compose.ts

I think we can use middleware composition this way with correct typing. But now it not applicable to HttpRequest.

What do you think about this guys?

**Comment by JozefFlakus** [comment]
Hi @RomanValihura!

Yes, this is an important topic that some time ago I discussed with @krzysztof-miemiec. I think it is possible to do and have some draft ideas how can we achieve it.

Basically `HttpRequest` interface should be generic and be able to define types for all possible kind of dynamic parameters like: `req.params`, `req.query`, `req.body`, eg:

```typescript
interface HttpRequest<Body = any, Query = any, Params = any> {
  ...
}
```

Having this, each middleware should return the observable of generic HttpRequest where the corresponding type is inferred inside the middleware implementation. The last thing, is to pass this types to `use` operator and properly infer it inside observable pipeline.

The only one problem that I can se right now is that we cannot infer the `HttpRequest` type inside `combineRoutes` function, but we can also think about ability to set generic type inside `EffectFactory` 🤔 instead. 

tl;dr I think this is a good topic for longer discussion how can we achieve this feature.

BTW. Thanks for kind words! It encourages us to work harder and deliver more features. 💪

**Comment by JozefFlakus** [comment]
@RomanValihura With the release of version 2.0, I'm closing this issue and marking it as **done**.
For more complex validation type inference I recommend using the brand new `@marblejs/middleware-io` middleware, based on @gcanti `io-ts` library. 🚀

**Comment by RomanValihura** [comment]
@JozefFlakus Awesome job 🎉
